src/consensus/tx_verify.cpp

```diff
@@ -181,25 +185,167 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state, bool fChe
     }
 
     // Check for duplicate inputs - note that this check is slow so we skip it in CheckBlock
-    if (fCheckDuplicateInputs) {
-        std::set<COutPoint> vInOutPoints;
-        for (const auto& txin : tx.vin)
-        {
-            if (!vInOutPoints.insert(txin.prevout).second)
-                return state.DoS(100, false, REJECT_INVALID, "bad-txns-inputs-duplicate");
-        }
-    }
-
-    if (tx.IsCoinBase())
-    {
+    if (tx.IsCoinBase() ) {
         if (tx.vin[0].scriptSig.size() < 2 || tx.vin[0].scriptSig.size() > 100)
             return state.DoS(100, false, REJECT_INVALID, "bad-cb-length");
-    }
-    else
-    {
-        for (const auto& txin : tx.vin)
-            if (txin.prevout.IsNull())
+    } else if (!fCheckDuplicateInputs || tx.vin.size() == 1){
+        for (const auto& txin : tx.vin) {
+            if (txin.prevout.IsNull()) {
                 return state.DoS(10, false, REJECT_INVALID, "bad-txns-prevout-null");
+            }
+        }
+    } else{
+        // This duplication checking algorithm uses a probabilistic filter
+        // to check for collisions efficiently.
+        //
+        // This is faster than the naive construction, using a set, which
+        // requires more allocation and comparison of uint256s.
+        //
+        // First we create a bitset table with 1<<21 elements. This
+        // is around 300 KB, so we construct it on the heap.
+        //
+        // We also allow reusing a 'dirty' table because zeroing 300 KB
+        // can be expensive, and the table will operate acceptably for all of the
+        // transactions in a given block.
+        //
+        // Then, we iterate through the elements one by one, generated 8 salted
```

Comment:
Why 8?

Reply:
It's my favorite number!

It's to get the expected work below one for the worst case as shown in the analysis.

We could add a 9th hash (we have more bits in the hash computed) but every hash adds more memory accesses in our table.

We could also remove a hash or two and the EV would be less than 10, which is probably acceptable 2. At about 4 or 5 hashes is when it blows up a bit more to an "unacceptable point" (EV 50 to 500 comparisons).

Solve for x such that:

Sum  i = 0 to 24390 [ i*( i*x / 2**21)**x ] < 1


It's also possible to modify the algorithm such that if a false positive is hit, you do the current set based algorithm up to the conflict. I'm not sure how to analyze that though, and the current code is sufficiently simple with low enough probability of expensive scan that we don't care that much.

Reply:
More precisely, it's one of two choices (9 hashes or 8) with 21 bits with expected work less than 1.

Extracting 9 hashes from 3 64 bit integers is a bit more complex code wise, but doable.

>>> sorted((sum(i*(float(i)*x/ 2.0**y)**x for i in xrange(24930)) if (x*y <= 192) else  ('Inf', 0, 0),y,x) for y in xrange(1,22) for x in xrange(1,20))[:10]

[(0.10374566662377155, 21, 9), (0.4157347268068221, 21, 8), (1.9074647424172138, 21, 7), (10.226961125517702, 21, 6), (53.11778131137103, 20, 9), (65.8563129753341, 21, 5), (106.42809006254646, 20, 8), (244.15548702940336, 20, 7), (529.481130078109, 21, 4), (654.5255120331329, 20, 6)]

Another option would be to increase the number of hashes to 16 (requiring a 256-bit hash) and then use a 20 bit table. This makes the expected work about 7 comparisons in the worst case, but makes the table half as large which reduces the constant bloat.

Reply:
Ok cool. Thanks for the analysis. 8 seems like a pretty reasonable choice. I left a comment above, but was wondering how this bloom filter compares to native unordered set implementations in the stdlib.